pkg/kv/kvserver/kvserverbase/forced_error.go

```diff
@@ -102,6 +121,49 @@ func CheckForcedErr(
 		}
 	}
 
+	// Perform some extra validation for lease requests and lease transfers if
+	// the BelowRaftPreviousLeaseAssertions is set.
+
+	// The ProposerLease field was introduced in v24.3. In a mixed version state,
+	// 23.1/2 nodes will not set it. Moreover, it is only set when the
+	// BelowRaftPreviousLeaseAssertions is set on the proposer.  So the field may
+	// be empty.
+	if BelowRaftPreviousLeaseAssertions && (isLeaseRequest || isLeaseTransfer) && !proposerLease.Empty() {
+		s1, s2 := proposerLease.Sequence, replicaState.Lease.Sequence
+		switch {
+		case s2 < s1:
+			// Our sequence number is lower than the proposer's. We may be behind in
+			// applying leases, but there can be a difference of at-most 1. Otherwise,
+			// we missed an application.
+			if s1-s2 > 1 {
+				log.Fatalf(ctx,
+					"lease sequence numbers differ by more than 1; our lease: %s; proposer's lease: %s",
+					replicaState.Lease, proposerLease,
+				)
+			}
+		case s2 == s1:
```

Comment:
When talking about this with @pav-kv and @nvanbenschoten yesterday, we concluded that we can't make any strong assertions about replicas requesting a lease. Like you pointed out on that PR, replicas should be forgiven for requesting a lease while being behind on their log (which will never apply). Nathan pointed out that two replicas could request an expired lease, of which one will win the race, and the second could have triggered the assertion on that PR. 

With the insight that we can't make any strong claims about leases being proposed, I think we can make claims about leases under which a lease proposal was made. We know that this lease was applied on the replica that is now proposing a lease. So this assertion is trying to say that two replicas can't apply a lease, with the same sequence number, but different leaseholders.  

<details> <summary> This would've triggered on all nodes in the lead up to the lease divergence issue we're seeing with our educated guess about the history. 
</summary>
```
On n25:
- seq=4224, epo=22, holder=n25 
- seq=4225, expiration based, holder=n30 (lease transfer)

On n30:
- seq=4224, epo=14, holder=n30 
- seq=4225, epo=15, holder=n30 
```
<details>

Reply:
👍🏽 makes sense